Rewrite program and improve performance

Juici opened the pull request:
*todo*

Juici committed:
Start a rewrite of the program

Split the crate into a library and binary.
Attempt to minimise allocations.

Juici committed:
Improve performance

Use O(log n) time complexity approach with memoization.

Juici committed:
Manually handle caching

Juici committed:
Fix parsing of digits

dcxSt commented:
Whoa, this is awesome; you're building it!

Here are some thoughts:
- It seems like you're implementing `fib` recursively using an `O(log n)^2` algorithm (did I get the time complexity of that right?). 
- For computing a range, I guess you're going to calculate the first two numbers with `_fib` and then add them together for the rest of the range
- Is the plan for this crate only to be a CLI, or might you expand into a Fibonacci library? (currently, it seems like [fast-fibonacci](https://crates.io/crates/fast-fibonacci) only allows you to compute the terms modulo another term)
- This might be kind of extra / outside of the scope of this crate, but do you think you might
  - implement a Fibonacci approximation for large numbers using powers of phi + some mathematical wizardry to compute error terms
    - If the user provides a number that is too large, the program could approximate the first few terms and the order of magnitude
  - implement the modular Fibonacci thingy that [fast-fibonacci](https://crates.io/crates/fast-fibonacci) has
- It seems like you're doing just fine on your own, but would you be willing to pair on this or, if you find my suggestions interesting, discuss this over a zoom at some point? (I'm in EST)

Also, I'm new to rust (and not a super-advanced programmer); what does this mean?
```
    if n & 1 != 0 {
        let k = (n + 1) / 2;
```

Juici commented:
`n & 1 != 0` checks if `n` is odd.

Juici committed:
Further optimise computation

Switch to rug for big integers, this provides a large performance boost.
Optimise the fibonacci algorithm and minimise allocations.
